include/zephyr/drivers/firmware/scmi/util.h

```diff
@@ -0,0 +1,281 @@
+ *	needed because the protocol driver needs to work with the
+ *	SCMI API **and** the subsystem API.
+ *
+ * @param node_id protocol node identifier
+ * @param init_fn pointer to protocol's initialization function
+ * @param api pointer to protocol's subsystem API
+ * @param pm pointer to the protocol's power management resources
+ * @param data pointer to protocol's private data
+ * @param config pointer to protocol's private constant data
+ * @param level protocol initialization level
+ * @param prio protocol's priority within its initialization level
+ */
+#define DT_SCMI_PROTOCOL_DEFINE(node_id, init_fn, pm, data, config,		\
+				level, prio, api)				\
+	DT_SCMI_TRANSPORT_CHANNELS_DECLARE(node_id)				\
+	DT_SCMI_PROTOCOL_DATA_DEFINE(node_id, DT_REG_ADDR(node_id), data);	\
+	DEVICE_DT_DEFINE(node_id, init_fn, pm,					\
+			 &SCMI_PROTOCOL_NAME(DT_REG_ADDR(node_id)),		\
+					     config, level, prio, api)
+
+/**
+ * @brief Just like DT_SCMI_PROTOCOL_DEFINE(), but uses an instance
+ * of a `DT_DRV_COMPAT` compatible instead of a node identifier
+ *
+ * @param inst instance number
+ * @param ... other parameters as expected by DT_SCMI_PROTOCOL_DEFINE()
+ */
+#define DT_INST_SCMI_PROTOCOL_DEFINE(inst, init_fn, pm, data, config,		\
+				     level, prio, api)				\
+	DT_SCMI_PROTOCOL_DEFINE(DT_INST(inst, DT_DRV_COMPAT), init_fn, pm,	\
+				data, config, level, prio, api)
+
+/**
+ * @brief Define an SCMI protocol with no device
+ *
+ * Variant of DT_SCMI_PROTOCOL_DEFINE(), but no `struct device` is
+ * created and no initialization function is called during system
+ * initialization. This is useful for protocols that are not really
+ * part of a subsystem with an API (e.g: pinctrl).
+ *
+ * @param node_id protocol node identifier
+ * @param data protocol private data
+ */
+#define DT_SCMI_PROTOCOL_DEFINE_NODEV(node_id, data)				\
+	DT_SCMI_TRANSPORT_CHANNELS_DECLARE(node_id)				\
+	DT_SCMI_PROTOCOL_DATA_DEFINE(node_id, DT_REG_ADDR(node_id), data)
+
+/**
+ * @brief Create an SCMI message field
+ *
+ * Data might not necessarily be encoded in the first
+ * x bits of an SCMI message parameter/return value.
+ * This comes in handy when building said parameters/
+ * return values.
+ *
+ * @param x value to encode
+ * @param mask value to perform bitwise-and with `x`
+ * @param shift value to left-shift masked `x`
+ */
+#define SCMI_FIELD_MAKE(x, mask, shift)\
+	(((uint32_t)(x) & (mask)) << (shift))
+
+/**
+ * @brief SCMI protocol IDs
+ *
+ * Each SCMI protocol is identified by an ID. Each
+ * of these IDs needs to be in decimal since they
+ * might be used to build protocol and static channel
+ * names.
+ */
+#define SCMI_PROTOCOL_BASE 16
+#define SCMI_PROTOCOL_POWER_DOMAIN 17
+#define SCMI_PROTOCOL_SYSTEM 18
+#define SCMI_PROTOCOL_PERF 19
+#define SCMI_PROTOCOL_CLOCK 20
+#define SCMI_PROTOCOL_SENSOR 21
+#define SCMI_PROTOCOL_RESET_DOMAIN 22
+#define SCMI_PROTOCOL_VOLTAGE_DOMAIN 23
+#define SCMI_PROTOCOL_PCAP_MONITOR 24
+#define SCMI_PROTOCOL_PINCTRL 25
```

Comment:
May I know whether SCMI protocol IDs are unified between different vendors or platforms?

Reply:
Depends on the range. If we're talking about the 0x10 - 0x19 range then yes. For instance, 0x10 will always be base protocol and so on.